[DAGCombiner] Fix scalarizeExtractedBinOp for some SETCC cases

david-arm opened the pull request:
PR https://github.com/llvm/llvm-project/pull/118823 added a
DAG combine for extracting elements of a vector returned from
SETCC, however it doesn't correctly deal with the case where
the vector element type is not i1. In this case we have to
take account of the boolean contents, which are represented
differently between vectors and scalars. The code now
explicitly performs an inreg sign extend in order to get the
same result.

Fixes https://github.com/llvm/llvm-project/issues/121372

arsenm commented on llvm/test/CodeGen/AArch64/extract-vector-cmp.ll:
```diff
@@ -141,6 +142,24 @@ for.cond.cleanup:
+; CHECK-NEXT:    cmp w8, #1
+; CHECK-NEXT:    csetm w0, lt
+; CHECK-NEXT:    ret
+  %2 = icmp ule <4 x i32> %0, zeroinitializer
```
Use named values in tests

david-arm replied:
Done

arsenm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -22809,7 +22809,8 @@ static SDValue scalarizeExtractedBinOp(SDNode *ExtElt, SelectionDAG &DAG,
   EVT ResVT = ExtElt->getValueType(0);
   if (Opc == ISD::SETCC &&
-      (ResVT != Vec.getValueType().getVectorElementType() || LegalTypes))
+      (ResVT != Vec.getValueType().getVectorElementType() || ResVT != MVT::i1 ||
```
Just check TLI.getBooleanContents?

david-arm replied:
Rather than bail out, I've just added support for detecting the additional sign-extend required.

david-arm committed:
Add test

david-arm committed:
[DAGCombiner] Fix issue #121372

PR #118823 added a DAG combine for extracting elements of a
vector returned from SETCC, however it doesn't correctly deal
with the case where the vector element type is not i1. In
this case we have to take account of the boolean contents,
which are represent differently between vectors and scalars.
The code now explicitly performs an inreg sign extend in
order to get the same result.

Fixes #121372

davemgreen commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -22808,9 +22808,26 @@ static SDValue scalarizeExtractedBinOp(SDNode *ExtElt, SelectionDAG &DAG,
+      return SDValue();
+
+    if (ResVT != MVT::i1) {
+      bool VecRequiresSignExt = TLI.getBooleanContents(VecVT) ==
```
I think it is worth making sure all cases work, in case a target with other BooleanContent runs into this in the future. I think the rule boils down to:
 - If you are converting to ZeroOrNegativeOneBooleanContent and from UndefinedBooleanContent or ZeroOrOneBooleanContent, it needs a sext_in_reg.
 - If you are converting to ZeroOrOneBooleanContent and from UndefinedBooleanContent or ZeroOrNegativeOneBooleanContent, it needs a zext_in_reg (an and).
 - Converting to UndefinedBooleanContent is always fine.

If you wanted to rule out the others (return SDValue) that sounds OK, considering the target needs to opt in. I think it would just be a case of tightening up what happens with the other cases.

david-arm replied:
Yeah I originally did that, but it's impossible to write a test case for the and (zext-in-reg) so the code will remain undefended. At least I couldn't find a target that exhibited that behaviour. I thought perhaps in this case it was better to bail out until a target required it? Having said that, I'm happy to support the case if we believe that testing isn't required.

david-arm committed:
Address review comments

davemgreen reviewed (approved):
Sorry I didn't get an update from github. LGTM.
